File: backend/tools/web_scraper_api.py
import httpx
import os
import re

BRIGHT_DATA_API_KEY = os.getenv("BRIGHT_DATA_API_KEY", "")
SCRAPER_BASE = "https://api.brightdata.com/datasets/v3"
LINKEDIN_DATASET = "gd_l1vikfnt1wgvvqz95w"
TIMEOUT = 25.0


async def enrich_company(company: str) -> dict:
    """Fetch structured company profile from Bright Data Web Scraper API."""
    if _is_linkedin_url(company):
        return await _enrich_via_linkedin(company)
    return await _enrich_via_web_unlocker(company)


def _is_linkedin_url(s: str) -> bool:
    return "linkedin.com/company" in s.lower()


async def _enrich_via_linkedin(url: str) -> dict:
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        trigger = await client.post(
            f"{SCRAPER_BASE}/trigger",
            params={"dataset_id": LINKEDIN_DATASET, "type": "discover_new", "discover_by": "url"},
            headers={
                "Authorization": f"Bearer {BRIGHT_DATA_API_KEY}",
                "Content-Type": "application/json",
            },
            json=[{"url": url}],
        )
        trigger.raise_for_status()
        data = trigger.json()

    snapshot_id = data.get("snapshot_id", "")
    if not snapshot_id:
        return _empty_profile(url)

    async with httpx.AsyncClient(timeout=30.0) as client:
        result = await client.get(
            f"{SCRAPER_BASE}/snapshot/{snapshot_id}",
            headers={"Authorization": f"Bearer {BRIGHT_DATA_API_KEY}"},
        )
        if result.status_code == 200:
            items = result.json()
            if items:
                return _parse_linkedin_profile(items[0], url)

    return _empty_profile(url)


async def _enrich_via_web_unlocker(company: str) -> dict:
    from tools.web_unlocker import scrape_with_unlocker
    from tools.content_extractor import extract_competitor_intelligence
    from bs4 import BeautifulSoup

    search_url = f"https://www.google.com/search?q={company.replace(' ', '+')}+company+site:linkedin.com"
    try:
        html = await scrape_with_unlocker(f"https://{company.replace(' ', '')}.com")
        soup = BeautifulSoup(html, "lxml")

        description = ""
        meta = soup.find("meta", attrs={"name": "description"})
        if meta:
            description = meta.get("content", "")

        return {
            "name": company,
            "industry": _guess_industry(html),
            "size_range": None,
            "headquarters": None,
            "website": f"https://{company.replace(' ', '').lower()}.com",
            "recent_news": [],
            "tech_signals": _extract_tech_signals(html),
            "linkedin_followers": None,
            "description": description,
        }
    except Exception:
        return _empty_profile(company)


def _parse_linkedin_profile(item: dict, url: str) -> dict:
    return {
        "name": item.get("name", ""),
        "industry": item.get("industry"),
        "size_range": item.get("company_size"),
        "headquarters": item.get("headquarters"),
        "website": item.get("website"),
        "recent_news": [p.get("text", "") for p in item.get("updates", [])[:3]],
        "tech_signals": [],
        "linkedin_followers": item.get("followers"),
    }


def _empty_profile(company: str) -> dict:
    return {
        "name": company,
        "industry": None,
        "size_range": None,
        "headquarters": None,
        "website": None,
        "recent_news": [],
        "tech_signals": [],
        "linkedin_followers": None,
    }


def _guess_industry(html: str) -> str | None:
    industries = ["software", "fintech", "healthcare", "e-commerce", "saas", "ai", "cloud", "security"]
    html_lower = html.lower()
    for ind in industries:
        if ind in html_lower:
            return ind.capitalize()
    return None


def _extract_tech_signals(html: str) -> list[str]:
    known_tech = ["React", "Python", "AWS", "GCP", "Azure", "Kubernetes", "Postgres", "Stripe", "Salesforce"]
    return [t for t in known_tech if t.lower() in html.lower()]

File: frontend/lib/tools/web-scraper-api.ts
import { scrapeWithUnlocker } from './web-unlocker';

const BRIGHT_DATA_API_KEY = process.env.BRIGHT_DATA_API_KEY ?? '';
const SCRAPER_BASE = 'https://api.brightdata.com/datasets/v3';
const LINKEDIN_DATASET = 'gd_l1vikfnt1wgvvqz95w';

const KNOWN_TECH = ['React', 'Python', 'AWS', 'GCP', 'Azure', 'Kubernetes', 'Postgres', 'Stripe', 'Salesforce'];
const INDUSTRIES = ['software', 'fintech', 'healthcare', 'e-commerce', 'saas', 'ai', 'cloud', 'security'];

export interface CompanyRaw {
  name: string; industry: string | null; size_range: string | null;
  headquarters: string | null; website: string | null;
  recent_news: string[]; tech_signals: string[]; linkedin_followers: number | null;
}

export async function enrichCompany(company: string): Promise<CompanyRaw> {
  if (company.toLowerCase().includes('linkedin.com/company')) {
    return enrichViaLinkedIn(company);
  }
  return enrichViaWebUnlocker(company);
}

async function enrichViaLinkedIn(url: string): Promise<CompanyRaw> {
  if (!BRIGHT_DATA_API_KEY) return emptyProfile(url);
  try {
    const trigger = await fetch(`${SCRAPER_BASE}/trigger?dataset_id=${LINKEDIN_DATASET}&type=discover_new&discover_by=url`, {
      method: 'POST',
      headers: { 'Authorization': `Bearer ${BRIGHT_DATA_API_KEY}`, 'Content-Type': 'application/json' },
      body: JSON.stringify([{ url }]),
      signal: AbortSignal.timeout(25_000),
    });
    if (!trigger.ok) return emptyProfile(url);
    const data = await trigger.json();
    const snapshotId = data?.snapshot_id;
    if (!snapshotId) return emptyProfile(url);

    const result = await fetch(`${SCRAPER_BASE}/snapshot/${snapshotId}`, {
      headers: { 'Authorization': `Bearer ${BRIGHT_DATA_API_KEY}` },
      signal: AbortSignal.timeout(30_000),
    });
    if (result.ok) {
      const items = await result.json();
      if (items?.[0]) return parseLinkedIn(items[0], url);
    }
  } catch {}
  return emptyProfile(url);
}

async function enrichViaWebUnlocker(company: string): Promise<CompanyRaw> {
  try {
    const domain = company.replace(/\s+/g, '').toLowerCase();
    const html = await scrapeWithUnlocker(`https://${domain}.com`);
    return {
      name: company,
      industry: guessIndustry(html),
      size_range: null,
      headquarters: null,
      website: `https://${domain}.com`,
      recent_news: [],
      tech_signals: extractTechSignals(html),
      linkedin_followers: null,
    };
  } catch {
    return emptyProfile(company);
  }
}

function parseLinkedIn(item: Record<string, unknown>, url: string): CompanyRaw {
  return {
    name: String(item.name ?? url),
    industry: item.industry ? String(item.industry) : null,
    size_range: item.company_size ? String(item.company_size) : null,
    headquarters: item.headquarters ? String(item.headquarters) : null,
    website: item.website ? String(item.website) : null,
    recent_news: ((item.updates as Record<string, string>[]) ?? []).slice(0, 3).map(p => p.text ?? ''),
    tech_signals: [],
    linkedin_followers: item.followers ? Number(item.followers) : null,
  };
}

function emptyProfile(name: string): CompanyRaw {
  return { name, industry: null, size_range: null, headquarters: null, website: null, recent_news: [], tech_signals: [], linkedin_followers: null };
}

function guessIndustry(html: string): string | null {
  const lower = html.toLowerCase();
  return INDUSTRIES.find(i => lower.includes(i))?.charAt(0).toUpperCase() + (INDUSTRIES.find(i => lower.includes(i))?.slice(1) ?? '') || null;
}

function extractTechSignals(html: string): string[] {
  return KNOWN_TECH.filter(t => html.toLowerCase().includes(t.toLowerCase()));
}
